libpod/image/pull.go

```diff
@@ -198,12 +200,74 @@ func (ir *Runtime) pullGoalFromImageReference(ctx context.Context, srcRef types.
 
 // pullImageFromHeuristicSource pulls an image based on inputName, which is heuristically parsed and may involve configured registries.
 // Use pullImageFromReference if the source is known precisely.
-func (ir *Runtime) pullImageFromHeuristicSource(ctx context.Context, inputName string, writer io.Writer, authfile, signaturePolicyPath string, signingOptions SigningOptions, dockerOptions *DockerRegistryOptions, label *string) ([]string, error) {
+func (ir *Runtime) pullImageFromHeuristicSource(ctx context.Context, config *NewImageConfig) ([]string, error) {
 	span, _ := opentracing.StartSpanFromContext(ctx, "pullImageFromHeuristicSource")
 	defer span.Finish()
+	sc := GetSystemContext(config.SignaturePolicyPath, config.Authfile, false)
+	goal, err := ir.getPullGoalFromHeuristicSource(ctx, sc, config.Name)
+	if err != nil {
+		return nil, err
+	}
+	return ir.doPullImage(ctx, sc, *goal, config.Writer, config.SigningOptions, config.DockerOptions, config.Label)
+}
+
+// pullImageFromHeuristicSourceIfNecessary checks an image is out of date with its source,
+// and if so, pulls an updated image.
+// Rather than search for the name in config, use the names of localImg, so we don't have to
+// refind the transport and reconstruct the correct name.
+func (ir *Runtime) pullImageFromHeuristicSourceIfNecessary(ctx context.Context, localImg *storage.Image, config *NewImageConfig) ([]string, error) {
+	span, _ := opentracing.StartSpanFromContext(ctx, "pullImageFromHeuristicSourceIfNecessary")
+	defer span.Finish()
+
+	var inspectErrors *multierror.Error
 
+	sc := GetSystemContext(config.SignaturePolicyPath, config.Authfile, false)
+	var goal *pullGoal
+	for _, name := range localImg.Names {
+		goal, _ = ir.getPullGoalFromHeuristicSource(ctx, sc, name)
+		if goal != nil {
+			break
+		}
+	}
+	if goal == nil {
+		return nil, errors.Errorf("Unable to get pull goal for updating image")
+	}
+
+	found := false
+	for _, imageInfo := range goal.refPairs {
+		if config.Writer != nil && (imageInfo.srcRef.Transport().Name() == DockerTransport || imageInfo.srcRef.Transport().Name() == AtomicTransport) {
+			logrus.Debugf("Checking if image is up to date with %s", imageInfo.image)
+		}
+
+		upToDate, err := checkRemoteImageIsUpToDate(ctx, config.Label, localImg.Created, imageInfo, sc)
+		if err != nil {
+			inspectErrors = multierror.Append(inspectErrors, err)
+			continue
+		}
+		if upToDate {
+			// we don't need to update our image.
+			return nil, nil
+		}
+		// Image was found and needs updating, break to pull the image
+		if config.Writer != nil {
+			logrus.Debugf("%s is not up to date with remote image, updating\n", config.Name)
+		}
+		found = true
+		break
+	}
+	if !found {
+		return nil, errors.Wrapf(inspectErrors, "%s was not found in any registry", config.Name)
```

Comment:
The way `Image.Names` works (see above), we are not using the search list, and `len(goal.RefPairs) == 1` here. So, “not found in any registry” is inaccurate.

(<s>OTOH it might make sense to embed this check inside the `range localImg.Names` loop, in case the same image is pulled from multiple registries over time, but the first one is down?</s>

No, this should all be based on users’ input.  If the users’ input specifies a particular registry, we should not be pulling an “updated” version of an image from an entirely different registry just because before the pull, the two registries pointed at exactly the same image.

So, the approach of looping over `localImg.Names` is fundamentally wrong; this should start with the user input, process the search etc. in the same way as other user input, and then perhaps match the `pullGoal.refPairs[].dstRef` against images existing locally.

Reply:
To be clear about the concern, consider:
```sh
author$ # create product:1.0
author$ podman push image:1.0 registryA.example.com/product:latest
author$ podman push image:1.0 registryB.example.com/product:latest

user$ podman pull branch.example.com/product:latest
user$ podman pull mainline.example.com/product:latest
# There is now a single local image with both names in image.Names

author$ # create product:1.1 and product:2.0
author$ podman push product:1.1 branch.example.com/product:latest
author$ podman push product:2.0 mainline.example.com/product:latest
# The registries have diverged

user$ podman pull --update branch.example.com/product:latest
# depending on whether branch.… or mainline.… is first in image.Names, this might
# update the image from the wrong branch, and update the wrong local tag!
```